drivers/serial/usart_gd32.c

```diff
@@ -93,25 +148,201 @@ static int usart_gd32_err_check(const struct device *dev)
+int usart_gd32_irq_rx_ready(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return usart_flag_get(cfg->reg, USART_FLAG_RBNE);
+}
+
+/** Interrupt driven error enabling function */
+void usart_gd32_irq_err_enable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_enable(cfg->reg, USART_INT_ERR);
+	usart_interrupt_enable(cfg->reg, USART_INT_PERR);
+}
+
+/** Interrupt driven error disabling function */
+void usart_gd32_irq_err_disable(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	usart_interrupt_disable(cfg->reg, USART_INT_ERR);
+	usart_interrupt_disable(cfg->reg, USART_INT_PERR);
+}
+
+/** Interrupt driven pending status function */
+int usart_gd32_irq_is_pending(const struct device *dev)
+{
+	const struct gd32_usart_config *const cfg = dev->config;
+
+	return (usart_flag_get(cfg->reg, USART_FLAG_RBNE) &&
+		usart_interrupt_flag_get(cfg->reg, USART_INT_RBNE)) ||
+		(usart_flag_get(cfg->reg, USART_FLAG_TC) &&
+		usart_interrupt_flag_get(cfg->reg, USART_INT_TC));
+}
+
+/** Interrupt driven interrupt update function */
+int usart_gd32_irq_update(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return 1;
+}
+
+/** Set the irq callback function */
+void usart_gd32_irq_callback_set(const struct device *dev,
+				uart_irq_callback_user_data_t cb,
+				void *user_data)
+{
+	struct gd32_usart_data *const data = dev->data;
+
+	data->user_cb = cb;
+	data->user_data = user_data;
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
 static const struct uart_driver_api usart_gd32_driver_api = {
 	.poll_in = usart_gd32_poll_in,
 	.poll_out = usart_gd32_poll_out,
 	.err_check = usart_gd32_err_check,
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_fill = usart_gd32_fifo_fill,
+	.fifo_read = usart_gd32_fifo_read,
+	.irq_tx_enable = usart_gd32_irq_tx_enable,
+	.irq_tx_disable = usart_gd32_irq_tx_disable,
+	.irq_tx_ready = usart_gd32_irq_tx_ready,
+	.irq_tx_complete = usart_gd32_irq_tx_complete,
+	.irq_rx_enable = usart_gd32_irq_rx_enable,
+	.irq_rx_disable = usart_gd32_irq_rx_disable,
+	.irq_rx_ready = usart_gd32_irq_rx_ready,
+	.irq_err_enable = usart_gd32_irq_err_enable,
+	.irq_err_disable = usart_gd32_irq_err_disable,
+	.irq_is_pending = usart_gd32_irq_is_pending,
+	.irq_update = usart_gd32_irq_update,
+	.irq_callback_set = usart_gd32_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 };
 
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+#define GD32_USART_IRQ_HANDLER_DECL(n)				\
+	static void usart_gd32_config_func_##n(const struct device *dev)
+#define GD32_USART_IRQ_HANDLER_FUNC(n)				\
+	.irq_config_func = usart_gd32_config_func_##n
+#define GD32_USART_IRQ_HANDLER(n)				\
+	static void usart_gd32_config_func_##n(const struct device *dev)			\
+	{								       \
+		IRQ_CONNECT(DT_INST_IRQN(n),				\
+			DT_INST_IRQ(n, priority),				\
+			usart_gd32_isr, DEVICE_DT_INST_GET(n),		\
+			0);							\
```

Comment:
`DT_INST_IRQ(n, priority)` should be vertically align with `DT_INST_IRQN(n),`.
Keep 1 parameter per line

```
usart_gd32_isr,
DEVICE_DT_INST_GET(n),
```

Reply:
Done.